torchdata/datapipes/iter/util/prefetch.py

```diff
@@ -0,0 +1,209 @@
+# Copyright (c) Meta Platforms, Inc. and affiliates.
+# All rights reserved.
+#
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree.
+
+import threading
+
+from collections import deque
+from concurrent.futures import Future, ThreadPoolExecutor, TimeoutError
+from dataclasses import dataclass
+from functools import partial
+from typing import Callable, Deque, Iterator, Optional, TypeVar
+
+import torch
+import torch.distributed as dist
+
+from torchdata._constants import default_timeout_in_s
+from torchdata.datapipes import functional_datapipe
+from torchdata.datapipes.iter import IterDataPipe
+
+T_co = TypeVar("T_co", covariant=True)
+
+
+__all__ = ["Expected", "FullSyncIterDataPipe", "PrefetchTimeoutError"]
+
+
+class PrefetchTimeoutError(RuntimeError):
+    def __init__(self, timeout: int) -> None:
+        super().__init__(f"Fail to fetch data within {timeout} seconds")
+
+
+class _EndOfPrefetch:
+    ...
+
+
+@dataclass
+class Expected:
+    r"""
+    Expected data provided to callback function in ``_PrefetchExecutor``.
+    """
+    index: int
+    error: Optional[BaseException] = None
+
+    def has_error(self) -> bool:
+        return self.error is not None
+
+
+class _PrefetchExecutor:
+    def __init__(
+        self,
+        datapipe_iterator: Iterator,
+        prefetch_size: int = 1,
+        callback_fn: Optional[Callable[[Expected], None]] = None,
+        timeout: int = default_timeout_in_s,
+    ) -> None:
+        self.datapipe_iterator = datapipe_iterator
+        self.prefetch_size = prefetch_size
+        self.callback_fn = callback_fn
+        self.timeout = timeout
+        # Use max_workers as 1 to guarantee the order of data fetched from iterator
+        self._executor = ThreadPoolExecutor(max_workers=1)
+        self._futures: Deque[Future] = deque()
+        self._lock = threading.RLock()
+        self._end_flag = False
+        self._idx = 0
+        for _ in range(prefetch_size):
+            with self._lock:
+                if self._end_flag:
+                    break
+            fetch_future: Future = self._executor.submit(self.fetch_next)
+            fetch_future.add_done_callback(partial(self._done_callback_fn, self._idx))
+            self._futures.append(fetch_future)
+            with self._lock:
+                self._idx += 1
+
+    def fetch_next(self):
+        return next(self.datapipe_iterator)
+
+    def _done_callback_fn(self, index: int, f: Future):
+        if f.exception():
+            with self._lock:
+                self._end_flag = True
```

Comment:
Question: for any exception (including `StopIteration`), is this setting the `_end_flag = True` but is not re-raising the exception? It will then be handled by `_executor` and `callback_fn` if available?

Reply:
Yeah. It's callback in a separate thread to set flag to DataPipe. And exception of StopIteration should be re-raised when __iter__ is invoked from Datapipe